README.md
![MasterFastfile - Kin + Carta Create](./MasterFastfile.png)

# MasterFastfile
[![GitHub release](https://img.shields.io/github/release/theappbusiness/masterfastfile/all.svg)](https://github.com/theappbusiness/MasterFastfile/releases/latest)
[![License](https://img.shields.io/badge/license-MIT-blue.svg?style=flat)](https://github.com/theappbusiness/MasterFastfile/blob/master/LICENSE)

To setup your project to use the MasterFastfile, navigate to your project root and run the following command:

```shell
curl https://raw.githubusercontent.com/theappbusiness/MasterFastfile/main/setup.sh | sh
```

Then setup your environments using the `.env` files created for you. To complete the setup just add the missing variable values.

To use the MasterFastfile add the following command to your Fastfile:

```ruby
import_from_git(url: 'https://github.com/theappbusiness/MasterFastfile.git', branch: '4.0.0', path: 'Fastfile')
```
For more detailed instructions [see our wiki](https://github.com/theappbusiness/MasterFastfile/wiki)

## Available lanes

* `test`
  * runs tests
* `ui_test`
  * runs ui tests
* `deploy_to_app_center`
  * runs tests
  * sets build number
  * installs provisioning profiles (if any are included in your Fastlane directory)
  * adds build info to app icon
  * builds and archives project
  * generates changelog from git commits
  * uploads app to hockey
* `deploy_to_app_center_no_test`
  * sets build number
  * adds build info to app icon
  * installs provisioning profiles
  * builds and archives project
  * generates changelog from git commits
  * uploads app to hockey
* `deploy_to_test_flight` (How do I find my iTunes Connect team ID? [link](https://github.com/fastlane/fastlane/issues/4301#issuecomment-253461017))
  * runs tests
  * sets build number
  * installs provisioning profiles
  * adds build info to app icon
  * builds and archives project
  * uploads app to test flight
* `local_build`
  * optionally adds icon overlay: e.g. `fastlane local_build icon_overlay:true`
  * installs provisioning profiles
  * builds an ipa

## Troubleshoot
* `TAB_PRIMARY_TARGET` if you are using multiple target in your project and it stops building for that reason, just give in the name of the target which holds the normal build number of your project

## Custom Actions

* `install_provisioning_profiles`
  * Automatically installs any provisioning profiles included in your Fastlane directory

* `icon_overlay`
  * Adds overlay to app icon containing build information

For more detailed information on how to setup your project and environment please see our [wiki](https://github.com/theappbusiness/MasterFastfile/wiki)

## Code signing

From Xcode 9 and onwards you are required to provide an export options plist. The MasterFastfile uses this to its advantage to handle as much code signing stuff as it can on your behalf.

For the MasterFastfile to fully take advantage of your export options ensure that you have defined the following environment variables:

- `GYM_EXPORT_OPTIONS`: The path to your export options, see below for how to generate these (you can choose different export options for different environments this way).
- `FL_PROJECT_SIGNING_PROJECT_PATH`: The path to your main Xcode project (not your workspace).
- `FL_UPDATE_PLIST_APP_IDENTIFIER`: The bundle identifier you want your main app target to have (you can choose a different identifier for different environments).

Export options define which provisioning profiles to use for which targets, as well as stuff like the team and export type.
The easiest way to figure out what your export options plist should contain is to manually set up code signing in Xcode\*, archive your app and then click "export...". Once exported, in the directory Xcode creates you'll be able to find an `ExportOptions.plist` file.

The MasterFastfile uses whichever export options you have set for the `GYM_EXPORT_OPTIONS` environment variable to determine which provisioning profiles to use, as well as automatically detect which team should be used to code sign the app, and which export option to use.

\*It's recommended to do this at least once anyway. You'll know for sure whether you have your code signing set up correctly, which will help debug issues you might encounter while using Fastlane.

## Dependencies

### Test Center plugin
Add the test center plugin to your fastlane setup for the `ui_test` lane. Follow the guide [here](https://github.com/lyndsey-ferguson/fastlane-plugin-test_center)

### App Center plugin
Add the app center plugin to your fastlane setup for the app center lanes. Follow the guide [here](https://github.com/microsoft/fastlane-plugin-appcenter)

### Imagemagick
Install using brew
```shell
brew install imagemagick
```
### Ghostscript
Install using brew
```shell
brew install ghostscript
```

## License

This project is licensed under the terms of the MIT license. See the [LICENSE](LICENSE) file.


CHANGELOG.md
# CHANGELOG

## [6.0.0](https://github.com/theappbusiness/MasterFastfile/releases/tag/6.0.0)

- UI tests are now executed using `fastlane-plugin-test_center` - https://github.com/lyndsey-ferguson/fastlane-plugin-test_center.
- Added new (optional) environment variable `MULTI_SCAN_TRY_COUNT` to set the number of times a test is retried before it's considered a fail.
- `TAB_UI_TEST_DEVICES` now supports the setting of multiple devices, as the name implies. If setting multiple, these should be ':' separated.
- Added new (optional) environment variable `MULTI_SCAN_PARALLEL_WORKER_COUNT` to set the number of simulators for batched UI tests.
- Added new (optional) environment variable `TAB_UNIT_TEST_DEVICE` to be able to set a unit test device independently. The `device` scan parameter was used previously which conflicted with the use of `devices` by `multi-scan`.

## [5.0.1](https://github.com/theappbusiness/MasterFastfile/releases/tag/5.0.1)

- Fixes an issue that was causing hockey uploads to fail.

## [5.0.0](https://github.com/theappbusiness/MasterFastfile/releases/tag/5.0.0)

- `TAB_PRIMARY_TARGET` environment variable is now mandatory for projects using Icon Overlay.

- As of v 2.86.0, fastlane requires a target for get_version_number, used for icon badging.
See discussion here: fastlane/fastlane#12177. See PR #64 for details.
- Use intended Bundle ID environment setting for each action
- Distinguishing between `FL_UPDATE_PLIST_APP_IDENTIFIER` and `FL_UPDATE_APP_IDENTIFIER`. See PR #63 for details.

## [4.0.0](https://github.com/theappbusiness/MasterFastfile/releases/tag/4.0.0)
### Changed
- Multiple code-level changes for addressing Rubocop errors.
- Provfiles are no longer used, instead you should define export options.
- All provisioning profiles included in the `fastlane` directory now get installed automatically.
- Updated `setup.sh`.
- Fixed an issue with including a Hockey download link when messaging to Slack.
- Fixed an issue where using `get_current_version` with the latest version of Fastlane would cause a prompt for the target.

## [3.0.1](https://github.com/theappbusiness/MasterFastfile/releases/tag/3.0.1)
### Added
- Release badge and link pointing to latest release version

### Changed
- CHANGELOG format

## [3.0.0](https://github.com/theappbusiness/MasterFastfile/releases/tag/3.0.0)
### Changed
- If `FL_PROJECT_TEAM_ID` is not set, the team ID is extracted from the environment's `GYM_EXPORT_OPTIONS` plist.

### Removed
- Use of `TAB_EXPORT_METHOD`, instead extracting the export method from the environment's `GYM_EXPORT_OPTIONS` plist.

## [2.3.0](https://github.com/theappbusiness/MasterFastfile/releases/tag/2.3.0)
### Added
- Lane dedicated to running `ui_test` schemes

## [2.2.0](https://github.com/theappbusiness/MasterFastfile/releases/tag/2.2.0)
### Added
- Support for signing multiple target projects using Provfile's

## [2.1.0](https://github.com/theappbusiness/MasterFastfile/releases/tag/2.1.0)
### Added
- `deploy_to_test_flight` for building and uploading to iTunes Connect

## [2.0.0](https://github.com/theappbusiness/MasterFastfile/releases/tag/2.0.0)
### Added
- New enviroment variable `TAB_EXPORT_METHOD` needed for `gym`, possible values are:
	- "app-store"
	- "ad-hoc"
	- "development"
	- "enterprise" // If not set this is used by default

### Changed
- Renamed `hockey` to `deploy_to_hockey`
- Renamed `hockey_no_test` to `deploy_to_hockey_no_test`

## [1.6.0](https://github.com/theappbusiness/MasterFastfile/releases/tag/1.6.0)
### Added
- Support for updating the team ID if `FL_PROJECT_SIGNING_PROJECT_PATH` and  `FL_PROJECT_TEAM_ID` are set

## [1.5.0](https://github.com/theappbusiness/MasterFastfile/releases/tag/1.5.0)
### Added
- `local_build` for building ipas locally

## [1.4.0](https://github.com/theappbusiness/MasterFastfile/releases/tag/1.4.0)
### Added
- Support for updating `PRODUCT_BUNDLE_IDENTIFIER`

## [1.3.0](https://github.com/theappbusiness/MasterFastfile/releases/tag/1.3.0)
### Added
- Support for installing a provisioning profile

## [1.2.0](https://github.com/theappbusiness/MasterFastfile/releases/tag/1.2.0)
### Added
- Support for Xcode 8 provisioning profile specifier


setup.sh
#!/bin/bash

mkdir fastlane
cd fastlane
echo "app_identifier = ENV['FL_UPDATE_PLIST_APP_IDENTIFIER']" > Appfile
echo "fastlane_version \"2.86.0\"
import_from_git(url: 'https://github.com/theappbusiness/MasterFastfile.git', path: 'Fastfile')" > Fastfile

function make_default_env_file {
  cat > .env.default <<EOF
#This is your default environment file
#Set environment variables used in all builds here
#More information on available environment variables can be found here https://github.com/theappbusiness/MasterFastfile/wiki/Quick-simple-setup-using-TAB-defaults

FL_PROJECT_PROVISIONING_PROJECT_PATH="" #Path to project to update provisioning on
FL_PROJECT_PROVISIONING_PROFILE_FILE="" #Path to provisioning profile to use for signing
FL_PROJECT_PROVISIONING_PROFILE_TARGET_FILTER="" #matches name or type of a target
FL_PROJECT_PROVISIONING_PROFILE_BUILD_CONFIGURATION="" #Optional specifies configurations to apply to, default is all
FL_PROJECT_PROVISIONING_CODE_SIGN_IDENTITY="" #Code signing identity to use
FL_PROJECT_TEAM_ID="" #Project team id
GYM_EXPORT_METHOD="" #Export method i.e. app-store, ad-hoc, package, enterprise, development, developer-id

FL_PROJECT_SIGNING_PROJECT_PATH="" # Path to xcode project to sign
FL_UPDATE_PLIST_PATH="" #Path to Info.plist of application
FL_SLACK_CHANNEL= #Slack channel to post build results to

SCAN_DEVICE="iPhone SE (11.4)"
SCAN_SLACK_CHANNEL="" #Slack channel to post test run info to

TAB_PROVISIONING_PROFILE="" #The name of the provisioning profile to use.
TAB_PROVISIONING_PROFILE_PATH="" #Path to the provisioning profile to install.
TAB_PRIMARY_TARGET="" #Main app target name
TAB_XCODE_PATH="" #Path to required xcode
TAB_UI_TEST_DEVICES="iPhone X (11.4)"
TAB_UI_TEST_SCHEME="" #Scheme to use for running UITests
TAB_SLACK_WEBHOOK_URL="" #Slack webhook for posting build information

ICON_OVERLAY_ASSETS_BUNDLE="" #Path to .xcassets containing AppIcon. If this variable exists build info will be added to the app icon

APPCENTER_API_TOKEN="" #Used to upload builds to app center
APPCENTER_OWNER_NAME=""  # when you're on your organization's page in AppCenter, this is the part of the URL slug after orgs: https://appcenter.ms/orgs/<owner_name>/applications

ITUNES_CONNECT_USERNAME="" #iTunes Connect login (usually email address)
ITUNES_CONNECT_TEAM_ID="" #The ID of your iTunes Connect team if you're in multiple teams https://github.com/fastlane/fastlane/issues/4301#issuecomment-253461017
ITUNES_CONNECT_PROVIDER="" #The provider short name to be used with the iTMSTransporter to identify your team
EOF
}

function make_custom_env_file {
  env=$1
  cat > .env.$env <<EOF
#This is your ${env} environment file
#Set environment variables used in ${env} builds here
GYM_SCHEME="" #Scheme name
FL_UPDATE_PLIST_APP_IDENTIFIER="" #App Bundle Identifier
FL_UPDATE_PLIST_DISPLAY_NAME="" #Display name of app

APPCENTER_APP_NAME="" #Name of the application in app center
APPCENTER_DISTRIBUTE_DESTINATIONS="" #Distribution groups to give access to this app
APPCENTER_DISTRIBUTE_NOTIFY_TESTERS="" #Set to true to email testers about new build.
EOF
}

make_default_env_file
make_custom_env_file "test"
make_custom_env_file "uat"
make_custom_env_file "staging"
make_custom_env_file "prod"
